README
NAME
    `Device::BusPirate' - interact with a Bus Pirate device

DESCRIPTION
    This module allows a program to interact with a Bus Pirate hardware
    electronics debugging device, attached over a USB-emulated serial port.
    In the following description, the reader is assumed to be generally
    aware of the device and its capabilities. For more information about the
    Bus Pirate see:

      http://dangerousprototypes.com/docs/Bus_Pirate

    This module and its various component modules are based on Future,
    allowing either synchronous or asynchronous communication with the
    attached hardware device. For simple synchronous situations, the class
    may be used on its own, and any method that returns a `Future' instance
    should immediately call the `get' method on that instance to wait for
    and obtain the eventual result:

     my $spi = $pirate->enter_mode( "SPI" )->get;

     $spi->power( 1 )->get;
     my $input = $spi->writeread_cs( $output )->get;

    A truely-asynchronous program would be built using a subclass that
    overrides the basic `read' and `sleep' methods for some event loop or
    similar; these can then be chained using the `then' method:

     my $input = $pirate->enter_mode( "SPI" )
       ->then( sub {
          my ( $spi ) = @_;

          $spi->power( 1 )->then( sub {
             $spi->writeread_cs( $output );
          });
       });

CONSTRUCTOR
  $pirate = Device::BusPirate->new( %args )
    Returns a new `Device::BusPirate' instance to communicate with the given
    device. Takes the following named arguments:

    serial => STRING
        Path to the serial port device node the Bus Pirate is attached to.
        If not supplied, the `BUS_PIRATE' environment variable is used;
        falling back on a default of /dev/ttyUSB0.

    baud => INT
        Serial baud rate to communicate at. Normally it should not be
        necessary to change this from its default of `115200'.

METHODS
    The following methods documented with a trailing call to `->get' return
    Future instances.

  $pirate->sleep( $timeout )->get
    Returns a `Future' that will become ready after the given timeout (in
    seconds), unless it is cancelled first.

  @result = $pirate->enter_mutex( $code )->get
    Acts as a mutex lock, to ensure only one block of code runs at once.
    Calls to `enter_mutex' will be queued up; each `$code' block will only
    be invoked once the `Future' returned from the previous has completed.

    Mode implementations should use this method to guard complete wire-level
    transactions, ensuring that multiple concurrent ones will not collide
    with each other.

  $mode = $pirate->enter_mode( $modename )->get
    Switches the attached device into the given mode, and returns an object
    to represent that hardware mode to interact with. This will be an
    instance of a class depending on the given mode name.

    `BB'
        The bit-banging mode. Returns an instance of
        Device::BusPirate::Mode::BB.

    `I2C'
        The I2C mode. Returns an instance of Device::BusPirate::Mode::I2C.

    `SPI'
        The SPI mode. Returns an instance of Device::BusPirate::Mode::SPI.

    Once a mode object has been created, most of the interaction with the
    device would be done using that mode object, as it will have methods
    relating to the specifics of that hardware mode. See the classes listed
    above for more information.

  $chip = $pirate->mount_chip( $chipname, %opts )->get
    Constructs a "chip" object; a helper designed to communicate with some
    particular hardware device (usually a specific chip) attached to the Bus
    Pirate. This will be a subclass of Device::BusPirate::Chip, and will
    likely provide various methods specific to the operation of that
    particular device.

    `$chipname' should match the name declared by the chip module, and other
    options passed in `%opts' will be passed to its constructor.

  $pirate->start->get
    Starts binary IO mode on the Bus Pirate device, enabling the module to
    actually communicate with it. Normally it is not necessary to call this
    method explicitly as it will be done by the setup code of the mode
    object.

  $pirate->stop
    Stops binary IO mode on the Bus Pirate device and returns it to user
    terminal mode. It may be polite to perform this at the end of a program
    to return it to a mode that a user can interact with normally on a
    terminal.

TODO
    *   More modes - I2C, UART, 1-wire, raw-wire

    *   Documentation/examples/actual implementations of truely-async
        subclass.

    *   PWM, AUX frequency measurement and ADC support.

AUTHOR
    Paul Evans <leonerd@leonerd.org.uk>



Changes
Revision history for Device-BusPirate

0.06    2014/11/01 15:29:58
        [CHANGES]
         * Document and expose the $pirate->sleep method as user API
         * Added synopsis documentation to each Mode class

        [BUGFIXES]
         * Fix for timer cancellation; avoids memory/CPU leak
         * Fix for sleep-only await with no IO read pending
         * Don't worry about Chip:: subclasses that don't define a CHIP
           constant

0.05    2014/10/20 22:04:21
        [CHANGES]
         * Support I2C mode
         * Provide 'enter_mutex' method on main Device::BusPirate object;
           document its use by the SPI and I2C modes
         * Migrate power/pullup/aux peripheral methods to common Mode base

0.04    2014/09/27 17:03:35
        [CHANGES]
         * Added a small Tickit-driven UI for interacting with BBIO mode
         * Expanded Mode::BB - ability to write/read multiple pins at once
         * Default the Bus Pirate location from $BUS_PIRATE env. var.

0.03    2014/09/16 17:05:51
        [CHANGES]
         * Defined a model for attaching high-level 'Chip' interface modules
         * Renamed 'SPI' mode config parameters to match Bus Pirate terminlogy
         * Also support CPOL/CPHA and SPI mode numbers

0.02    2014/08/12 10:50:07
        [CHANGES]
         * Support bit-bang mode directly
         * Renamed SPI's pin_3v3 configuration to open_drain, as it's clearer
           in meaning
         * Better timeouts and non-blocking behaviour
         * Allow setting SPI bus speed
         * Allow SPI bulk transfers of more than 16 bytes
         * Rename 'SCK' to 'CLK' to match Bus Pirate terminology
         * Document that ->stop doesn't return a Future

0.01    2014/08/11 16:05:39
        First version, released on an unsuspecting world.



examples/bbio-tickit-ui.pl
#!/usr/bin/perl

use strict;
use warnings;

use Tickit;
use Tickit::Widgets qw( GridBox Button Static VBox CheckButton );
use Device::BusPirate;
use Getopt::Long;

my $PIRATE = "/dev/ttyUSB0";
my $BAUD = 115200;

GetOptions(
   'p|pirate=s' => \$PIRATE,
   'b|baud=i'   => \$BAUD,
) or exit 1;

my $pirate = Device::BusPirate->new(
   serial => $PIRATE,
   baud   => $BAUD,
);

my $bb = $pirate->enter_mode( "BB" )->get;

my $tickit = Tickit->new(
   root => my $grid = Tickit::Widget::GridBox->new(
      style => {
         row_spacing => 1,
         col_spacing => 2,
      },
      children => [
         [ Tickit::Widget::Static->new( text => "Line", heading => 1 ),
           ],
      ]
   )
);

Tickit::Style->load_style( <<'EOSTYLE' );
Static.heading {
  b: 1;
}

Static.miso { bg: "brown";  fg: "white"; }
Static.cs   { bg: "red";    fg: "white"; }
Static.mosi { bg: "orange"; fg: "black"; }
Static.clk  { bg: "yellow"; fg: "black"; }
Static.aux  { bg: "green";  fg: "black"; }

Button {
  bg: "black"; fg: "white";
}
Button:current {
  bg: "white"; fg: "black";
}

Button.miso:input { bg: "brown";  fg: "white"; }
Button.cs  :input { bg: "red";    fg: "white"; }
Button.mosi:input { bg: "orange"; fg: "black"; }
Button.clk :input { bg: "yellow"; fg: "black"; }
Button.aux :input { bg: "green";  fg: "black"; }

EOSTYLE

my %buttons;

my %pin_keys = (
   power => 'p',
   aux   => 'a',
   clk   => 'c',
   mosi  => 'o',
   cs    => 's',
   miso  => 'i'
);
foreach my $pin (qw( power aux clk mosi cs miso )) {
   my $row = $grid->rowcount;
   my $key = $pin_keys{$pin};

   $grid->add( $row, 0,
      Tickit::Widget::Static->new( text => "\U$pin\E\n\U$key\E / $key",
         class => $pin,
      ),
      row_expand => 1,
   );

   $grid->add( $row, 1,
      my $btn_on = $buttons{$pin}[1] = Tickit::Widget::Button->new( label => "on",
         class => $pin,
         on_click => sub { set_pin( $pin, 1 ) }
      ),
      col_expand => 1,
      row_expand => 1,
   );

   $grid->add( $row, 2,
       my $btn_off = $buttons{$pin}[0] = Tickit::Widget::Button->new( label => "off",
         class => $pin,
         on_click => sub { set_pin( $pin, 0 ) }
      ),
      col_expand => 1,
      row_expand => 1,
   );

   # Key bindings
   $tickit->bind_key( lc $key => sub { $btn_off->click } );
   $tickit->bind_key( uc $key => sub { $btn_on->click  } );

   if( $pin eq "power" ) {
      # Power can't be set to read mode
      $btn_off->set_style_tag( current => 1 );
      next;
   }

   $grid->add( $row, 3,
      my $btn_read = $buttons{$pin}[2] = Tickit::Widget::Button->new( label => "read",
         on_click => sub { set_pin_input( $pin ) }
      ),
      col_expand => 1,
      row_expand => 1,
   );

   # Initial states - all are read
   $btn_read->set_style_tag( current => 1 );
}

# Stick a couple of checkbuttons in the spare 'power' slot

$grid->add( 1, 3, Tickit::Widget::VBox->new(
      children => [
         Tickit::Widget::CheckButton->new(
            label => "Pullup",
            on_toggle => sub {
               my ( undef, $pullup ) = @_;
               $bb->pullup( $pullup )->get;
            },
         ),
         Tickit::Widget::CheckButton->new(
            label => "Open-drain",
            on_toggle => sub {
               my ( undef, $opendrain ) = @_;
               $bb->configure( open_drain => $opendrain )->get;
            },
         ),
      ],
) );

sub _tick
{
   my $pins = $bb->read->get;
   foreach my $pin ( keys %$pins ) {
      my $val = !!$pins->{$pin};
      $buttons{$pin}[$_]->set_style_tag( input => not( $val ^ $_ ) ) for 0 .. 1;
   }

   $tickit->timer( after => 0.05, \&_tick );
}

_tick;
$tickit->run;

END {
   $pirate and $pirate->stop;
}

sub set_pin
{
   my ( $pin, $val ) = @_;

   $bb->$pin( $val )->get;

   $buttons{$pin}[0]->set_style_tag( current =>  !$val );
   $buttons{$pin}[1]->set_style_tag( current => !!$val );
   $buttons{$pin}[2]->set_style_tag( current =>      0 ) if $buttons{$pin}[2];
   $buttons{$pin}[$_]->set_style_tag( input => 0 ) for 0 .. 1;
}

sub set_pin_input
{
   my ( $pin ) = @_;

   $bb->${\"read_$pin"}()->get; # Ignore it for now

   $buttons{$pin}[0]->set_style_tag( current => 0 );
   $buttons{$pin}[1]->set_style_tag( current => 0 );
   $buttons{$pin}[2]->set_style_tag( current => 1 );
}


lib/Device/BusPirate/Chip.pm
#  You may distribute under the terms of either the GNU General Public License
#  or the Artistic License (the same terms as Perl itself)
#
#  (C) Paul Evans, 2014 -- leonerd@leonerd.org.uk

package Device::BusPirate::Chip;

use strict;
use warnings;

our $VERSION = '0.06';

=head1 NAME

C<Device::BusPirate::Chip> - base class for chip-specific adapters

=head1 DESCRIPTION

This base class is provided for implementations of chip-specific adapters, to
provide specific code to interact with particular kinds of chip or other
hardware attached to the Bus Pirate.

=cut

=head1 CONSTRUCTOR

=cut

=head2 $chip = Device::BusPirate::Chip->new( $bp, %opts )

Constructs a new instance of the chip adapter. The first argument is the
L<Device::BusPirate> object itself. The remaining arguments are not otherwise
inspected by the base class; they are free for use by the specific subclasses.

=cut

sub new
{
   my $class = shift;
   my ( $bp ) = @_;

   return bless {
      bp => $bp,
   }, $class;
}

=head1 METHODS

=cut

=head2 $pirate = $chip->pirate

Returns the L<Device::BusPirate> instance.

=cut

sub pirate
{
   my $self = shift;
   return $self->{bp};
}

=head2 $mode = $chip->mode

Returns the C<Device::BusPirate::Mode> instance being used to communicate with
this chip. Normally this method would be used directly by the implementation
subclass, rather than the end-user code calling it.

=cut

sub mode
{
   my $self = shift;
   return $self->{mode};
}

=head2 $chip = $chip->mount( $mode )->get

Called by the C<mount_chip> method on the C<Device::BusPirate> object, this
method is intended for implementation subclasses to perform any initial
configuration of the C<$mode> instance that they require; such as details of
communication speed or settings.

=cut

sub mount
{
   my $self = shift;
   ( $self->{mode} ) = @_;
   Future->done( $self );
}

=head1 MODE PASSTHROUGH METHODS

The following methods are passed through to the C<mode> instance

 power
 pullup
 aux

=cut

foreach my $method (qw( power pullup aux )) {
   no strict 'refs';
   *$method = sub { shift->mode->$method( @_ ) };
}

=head1 IMPLEMENTATION METHODS

The following methods must be provided by specific implementations of this
base class:

=head2 $chipname = Device::BusPirate::Chip->CHIP

Returns the name for this chip; the name that must be passed to the
C<mount_chip> method on C<Device::BusPirate> in order to request this
particular implementation. For ease of use, this name should match the name
given by the hardware manufacturer to identify the chip. It does not need to
be a valid Perl symbol; it may contain characters not normally allowed in
symbol names.

=head2 $modename = $chip->MODE

Called after construction, this method should return a mode name that will be
used to communicate with the chip. Normally this would be a constant, hence
its capitalised name, but it is invoked on the instance, after construction,
in case of chips with multiple possible access methods (e.g. dual SPI or I2C
devices), allowing the method chance to inspect the constructor options to
make this choice.

=cut

=head1 AUTHOR

Paul Evans <leonerd@leonerd.org.uk>

=cut

0x55AA;


lib/Device/BusPirate/Mode.pm
#  You may distribute under the terms of either the GNU General Public License
#  or the Artistic License (the same terms as Perl itself)
#
#  (C) Paul Evans, 2014 -- leonerd@leonerd.org.uk

package Device::BusPirate::Mode;

use strict;
use warnings;

our $VERSION = '0.06';

use constant {
   CONF_CS     => 0x01,
   CONF_AUX    => 0x02,
   CONF_PULLUP => 0x04,
   CONF_POWER  => 0x08,
};

=head1 NAME

C<Device::BusPirate::Mode> - base class for C<Device::BusPirate> modes

=head1 DESCRIPTION

The following methods are implemented by all the various mode subclasses.

=cut

sub new
{
   my $class = shift;
   my ( $bp ) = @_;

   my $self = bless {
      bp => $bp,
   }, $class;

   $self->{cs}     = 0;
   $self->{power}  = 0;
   $self->{pullup} = 0;
   $self->{aux}    = 0;

   return $self;
}

=head1 METHODS

=cut

=head2 $pirate = $mode->pirate

Returns the underlying L<Device::BusPirate> instance.

=cut

sub pirate
{
   my $self = shift;
   return $self->{bp};
}

sub _start_mode_and_await
{
   my $self = shift;
   my ( $send, $await ) = @_;

   my $pirate = $self->pirate;

   $pirate->write( $send );
   $pirate->read( length $await )->then( sub {
      my ( $buf ) = @_;
      return Future->done( $buf ) if $buf eq $await;
      return Future->fail( "Expected '$await' response but got '$buf'" );
   });
}

=head2 $mode->power( $power )->get

Enable or disable the C<VREG> 5V and 3.3V power outputs.

=cut

sub power
{
   my $self = shift;
   $self->{power} = !!shift;
   $self->_update_peripherals;
}

=head2 $mode->pullup( $pullup )->get

Enable or disable the IO pin pullup resistors from C<Vpu>. These are connected
to the C<MISO>, C<CLK>, C<MOSI> and C<CS> pins.

=cut

sub pullup
{
   my $self = shift;
   $self->{pullup} = !!shift;
   $self->_update_peripherals;
}

=head2 $mode->aux( $aux )->get

Set the C<AUX> output pin level.

=cut

sub aux
{
   my $self = shift;
   $self->{aux} = !!shift;
   $self->_update_peripherals;
}

=head2 $mode->cs( $cs )->get

Set the C<CS> output pin level.

=cut

sub cs
{
   my $self = shift;
   $self->{cs} = !!shift;
   $self->_update_peripherals;
}

sub _update_peripherals
{
   my $self = shift;

   $self->pirate->write( chr( 0x40 |
      ( $self->{power}  ? CONF_POWER  : 0 ) |
      ( $self->{pullup} ? CONF_PULLUP : 0 ) |
      ( $self->{aux}    ? CONF_AUX    : 0 ) |
      ( $self->{cs}     ? CONF_CS     : 0 ) )
   );
   $self->pirate->read( 1 )
      ->then( sub {
         my( $buf ) = @_;
         $buf eq "\x01" or return Future->fail( "Expected ACK response to _update_peripherals" );
         return Future->done;
      });
}

=head1 AUTHOR

Paul Evans <leonerd@leonerd.org.uk>

=cut

0x55AA;


lib/Device/BusPirate/Mode/BB.pm
#  You may distribute under the terms of either the GNU General Public License
#  or the Artistic License (the same terms as Perl itself)
#
#  (C) Paul Evans, 2014 -- leonerd@leonerd.org.uk

package Device::BusPirate::Mode::BB;

use strict;
use warnings;
use base qw( Device::BusPirate::Mode );

our $VERSION = '0.06';

use Carp;

use Future;

use constant MODE => "BB";

use constant {
   MASK_CS   => 0x01,
   MASK_MISO => 0x02,
   MASK_CLK  => 0x04,
   MASK_MOSI => 0x08,
   MASK_AUX  => 0x10,

   CONF_PULLUP => 0x20,
   CONF_POWER  => 0x40,
};

# Convenience hash
my %PIN_MASK = map { $_ => __PACKAGE__->${\"MASK_\U$_"} } qw( cs miso clk mosi aux );

=head1 NAME

C<Device::BusPirate::Mode::BB> - use C<Device::BusPirate> in bit-banging mode

=head1 SYNOPSIS

 use Device::BusPirate;

 my $pirate = Device::BusPirate->new;
 my $bb = $pirate->enter_mode( "BB" )->get;

 my $count = 0;
 while(1) {
    $bb->write(
       miso => $count == 0,
       cs   => $count == 1,
       mosi => $count == 2,
       clk  => $count == 3,
       aux  => $count == 4,
    )->then( sub { $pirate->sleep( 0.5 ) })
     ->get;

   $count++;
   $count = 0 if $count >= 5;
 }

=head1 DESCRIPTION

This object is returned by a L<Device::BusPirate> instance when switching it
into C<BB> mode. It provides methods to configure the hardware, and interact
with the five basic IO lines in bit-banging mode.

=cut

=head1 METHODS

=cut

sub start
{
   my $self = shift;

   $self->{dir_mask} = 0x1f; # all inputs

   $self->{out_mask} = 0; # all off

   Future->done( $self );
}

=head2 $bb->configure( %args )->get

Change configuration options. The following options exist; all of which are
simple true/false booleans.

=over 4

=item open_drain

If enabled, a "high" output pin will be set as an input; i.e. hi-Z. When
disabled (default), a "high" output pin will be driven by 3.3V. A "low" output
will be driven to GND in either case.

=back

=cut

sub configure
{
   my $self = shift;
   my %args = @_;

   defined $args{$_} and $self->{$_} = $args{$_}
      for (qw( open_drain ));

   return Future->done
}

=head2 $bb->cs( $state )->get

=head2 $bb->miso( $state )->get

=head2 $bb->clk( $state )->get

=head2 $bb->mosi( $state )->get

=head2 $bb->aux( $state )->get

Set an output pin to the given logical state. Uses the C<open_drain>
configuration setting to determine whether high should be hi-Z or 3.3V.

=cut

sub cs   { shift->_write( 0, cs   => $_[0] ) }
sub miso { shift->_write( 0, miso => $_[0] ) }
sub clk  { shift->_write( 0, clk  => $_[0] ) }
sub mosi { shift->_write( 0, mosi => $_[0] ) }
sub aux  { shift->_write( 0, aux  => $_[0] ) }

=head2 $bb->write( %pins )->get

Sets the state of multiple output pins at the same time.

=cut

sub _write
{
   my $self = shift;
   my ( $want_read, %pins ) = @_;

   my $out = $self->{out_mask};
   my $dir = $self->{dir_mask};

   foreach my $pin ( keys %PIN_MASK ) {
      next unless exists $pins{$pin};
      my $mask = $PIN_MASK{$pin};

      if( $pins{$pin} and !$self->{open_drain} ) {
         $dir &= ~$mask;
         $out |=  $mask;
      }
      elsif( $pins{$pin} ) {
         $dir |=  $mask;
      }
      else {
         $dir &= ~$mask;
         $out &= ~$mask;
      }
   }

   my $len = 0;
   if( $dir != $self->{dir_mask} ) {
      $self->pirate->write( chr( 0x40 | $dir ) );
      $len++;

      $self->{dir_mask} = $dir;
   }

   if( $want_read or $out != $self->{out_mask} ) {
      $self->pirate->write( chr( 0x80 | $out ) );
      $len++;

      $self->{out_mask} = $out;
   }

   return Future->done unless $len;

   my $f = $self->pirate->read( $len );
   return $f if $want_read;
   return $f->then_done(); # ignore input
}

sub write
{
   my $self = shift;
   $self->_write( 0, @_ );
}

sub read_cs   { shift->_input1( MASK_CS,   @_ ) }
sub read_miso { shift->_input1( MASK_MISO, @_ ) }
sub read_clk  { shift->_input1( MASK_CLK,  @_ ) }
sub read_mosi { shift->_input1( MASK_MOSI, @_ ) }
sub read_aux  { shift->_input1( MASK_AUX,  @_ ) }

=head2 $state = $bb->read_cs->get

=head2 $state = $bb->read_miso->get

=head2 $state = $bb->read_clk->get

=head2 $state = $bb->read_mosi->get

=head2 $state = $bb->read_aux->get

Set a pin to input direction and read its current state.

=cut

sub _input1
{
   my $self = shift;
   my ( $mask ) = @_;

   $self->_input( $mask )->then( sub {
      my ( $buf ) = @_;
      Future->done( ord( $buf ) & $mask );
   });
}

sub _input
{
   my $self = shift;
   my ( $mask ) = @_;

   $self->{dir_mask} |= $mask;
   $self->pirate->write( chr( 0x40 | $self->{dir_mask} ) );
   return $self->pirate->read( 1 );
}

=head2 $pins = $bbio->read->get

Returns a HASH containing the current state of all the pins currently
configured as inputs. More efficient than calling multiple C<read_*> methods
when more than one pin is being read at the same time.

=cut

sub read
{
   my $self = shift;
   $self->writeread();
}

=head2 $in_pins = $bbio->writeread( %out_pins )->get

Combines the effects of C<write> and C<read> in a single operation; sets the
output state of any pins in C<%out_pins> then returns the input state of the
pins currently set as inputs.

=cut

sub writeread
{
   my $self = shift;
   $self->_write( 1, @_ )->then( sub {
      my ( $buf ) = @_;
      $buf = ord $buf;

      my $pins;
      foreach my $pin ( keys %PIN_MASK ) {
         my $mask = $PIN_MASK{$pin};
         next unless $self->{dir_mask} & $mask;
         $pins->{$pin} = !!( $buf & $mask );
      }
      Future->done( $pins );
   });
}

=head2 $bb->power( $power )->get

Enable or disable the C<VREG> 5V and 3.3V power outputs.

=cut

sub power
{
   my $self = shift;
   my ( $state ) = @_;

   $state ? ( $self->{out_mask} |=  CONF_POWER )
          : ( $self->{out_mask} &= ~CONF_POWER );
   $self->pirate->write( chr( 0x80 | $self->{out_mask} ) );
   $self->pirate->read( 1 )->then_done(); # ignore input
}

=head2 $bb->pullup( $pullup )->get

Enable or disable the IO pin pullup resistors from C<Vpu>. These are connected
to the C<MISO>, C<CLK>, C<MOSI> and C<CS> pins.

=cut

sub pullup
{
   my $self = shift;
   my ( $state ) = @_;

   $state ? ( $self->{out_mask} |=  CONF_PULLUP )
          : ( $self->{out_mask} &= ~CONF_PULLUP );
   $self->pirate->write( chr( 0x80 | $self->{out_mask} ) );
   $self->pirate->read( 1 )->then_done(); # ignore input
}

=head1 AUTHOR

Paul Evans <leonerd@leonerd.org.uk>

=cut

0x55AA;


lib/Device/BusPirate.pm
#  You may distribute under the terms of either the GNU General Public License
#  or the Artistic License (the same terms as Perl itself)
#
#  (C) Paul Evans, 2014 -- leonerd@leonerd.org.uk

package Device::BusPirate;

use strict;
use warnings;

our $VERSION = '0.06';

use Carp;

use Future::Utils qw( repeat );
use IO::Termios;
use Time::HiRes qw( time );

use Module::Pluggable
   search_path => "Device::BusPirate::Mode",
   require     => 1,
   sub_name    => "modes";
my %MODEMAP = map { $_->MODE => $_ } __PACKAGE__->modes;

use Module::Pluggable
   search_path => "Device::BusPirate::Chip",
   require     => 1,
   sub_name    => "chips";
my %CHIPMAP = map { $_->can( "CHIP" ) ? ( $_->CHIP => $_ ) : () } __PACKAGE__->chips;

use constant BUS_PIRATE => $ENV{BUS_PIRATE} || "/dev/ttyUSB0";
use constant PIRATE_DEBUG => $ENV{PIRATE_DEBUG};

=head1 NAME

C<Device::BusPirate> - interact with a F<Bus Pirate> device

=head1 DESCRIPTION

This module allows a program to interact with a F<Bus Pirate> hardware
electronics debugging device, attached over a USB-emulated serial port. In the
following description, the reader is assumed to be generally aware of the
device and its capabilities. For more information about the F<Bus Pirate> see:

=over 2

L<http://dangerousprototypes.com/docs/Bus_Pirate>

=back

This module and its various component modules are based on L<Future>, allowing
either synchronous or asynchronous communication with the attached hardware
device. For simple synchronous situations, the class may be used on its own,
and any method that returns a C<Future> instance should immediately call the
C<get> method on that instance to wait for and obtain the eventual result:

 my $spi = $pirate->enter_mode( "SPI" )->get;

 $spi->power( 1 )->get;
 my $input = $spi->writeread_cs( $output )->get;

A truely-asynchronous program would be built using a subclass that overrides
the basic C<read> and C<sleep> methods for some event loop or similar; these
can then be chained using the C<then> method:

 my $input = $pirate->enter_mode( "SPI" )
   ->then( sub {
      my ( $spi ) = @_;

      $spi->power( 1 )->then( sub {
         $spi->writeread_cs( $output );
      });
   });

=cut

=head1 CONSTRUCTOR

=cut

=head2 $pirate = Device::BusPirate->new( %args )

Returns a new C<Device::BusPirate> instance to communicate with the given
device. Takes the following named arguments:

=over 4

=item serial => STRING

Path to the serial port device node the Bus Pirate is attached to. If not
supplied, the C<BUS_PIRATE> environment variable is used; falling back on a
default of F</dev/ttyUSB0>.

=item baud => INT

Serial baud rate to communicate at. Normally it should not be necessary to
change this from its default of C<115200>.

=back

=cut

sub new
{
   my $class = shift;
   my %args = @_;

   my $serial = $args{serial} || BUS_PIRATE;
   my $baud   = $args{baud} || 115200;

   my $fh = IO::Termios->open( $serial, "$baud,8,n,1" )
      or croak "Cannot open serial port $serial - $!";

   $fh->setflag_icanon( 0 );
   $fh->setflag_echo( 0 );

   $fh->blocking( 0 );

   return bless {
      fh => $fh,
   }, $class;
}

=head1 METHODS

The following methods documented with a trailing call to C<< ->get >> return
L<Future> instances.

=cut

# For Modes
sub write
{
   my $self = shift;
   my ( $buf ) = @_;

   printf STDERR "PIRATE >> %v02x\n", $buf if PIRATE_DEBUG;

   $self->{fh}->syswrite( $buf );
}

# For Modes
sub read
{
   my $self = shift;
   my ( $n ) = @_;

   push @{ $self->{read_f} }, [ $n, my $f = $self->_new_future ];

   return $f;
}

=head2 $pirate->sleep( $timeout )->get

Returns a C<Future> that will become ready after the given timeout (in
seconds), unless it is cancelled first.

=cut

sub sleep
{
   my $self = shift;
   my ( $timeout ) = @_;

   my $until = time() + $timeout;

   croak "Cannot sleep less than existing timeout" if
      $self->{alarms}[0] and $until < $self->{alarms}[0][0];

   push @{ $self->{alarms} }, my $alarm = [ $until, my $f = $self->_new_future ];
   $f->on_cancel( sub {
      $self->{alarms} = [ grep { $_ != $alarm } @{ $self->{alarms} } ];
   });

   return $f;
}

=head2 @result = $pirate->enter_mutex( $code )->get

Acts as a mutex lock, to ensure only one block of code runs at once. Calls to
C<enter_mutex> will be queued up; each C<$code> block will only be invoked
once the C<Future> returned from the previous has completed.

Mode implementations should use this method to guard complete wire-level
transactions, ensuring that multiple concurrent ones will not collide with
each other.

=cut

sub enter_mutex
{
   my $self = shift;
   my ( $code ) = @_;

   my $oldm = $self->{mutex} // $self->_new_future->done( $self );
   $self->{mutex} = my $newm = $self->_new_future;

   $oldm->then( $code )
      ->then_with_f( sub {
         my $f = shift;
         $newm->done( $self );
         $f
      });
}

=head2 $mode = $pirate->enter_mode( $modename )->get

Switches the attached device into the given mode, and returns an object to
represent that hardware mode to interact with. This will be an instance of a
class depending on the given mode name.

=over 4

=item C<BB>

The bit-banging mode. Returns an instance of L<Device::BusPirate::Mode::BB>.

=item C<I2C>

The I2C mode. Returns an instance of L<Device::BusPirate::Mode::I2C>.

=item C<SPI>

The SPI mode. Returns an instance of L<Device::BusPirate::Mode::SPI>.

=back

Once a mode object has been created, most of the interaction with the device
would be done using that mode object, as it will have methods relating to the
specifics of that hardware mode. See the classes listed above for more
information.

=cut

sub enter_mode
{
   my $self = shift;
   my ( $modename ) = @_;

   my $modeclass = $MODEMAP{$modename} or
      croak "Unrecognised mode '$modename'";

   $self->start->then( sub {
      ( $self->{mode} = $modeclass->new( $self ) )->start;
   });
}

=head2 $chip = $pirate->mount_chip( $chipname, %opts )->get

Constructs a "chip" object; a helper designed to communicate with some
particular hardware device (usually a specific chip) attached to the Bus
Pirate. This will be a subclass of L<Device::BusPirate::Chip>, and will likely
provide various methods specific to the operation of that particular device.

C<$chipname> should match the name declared by the chip module, and other
options passed in C<%opts> will be passed to its constructor.

=cut

sub mount_chip
{
   my $self = shift;
   my ( $chipname, %opts ) = @_;

   my $chipclass = $CHIPMAP{$chipname} or
      croak "Unrecognised chip '$chipname'";

   my $chip = $self->{chip} = $chipclass->new( $self, %opts );

   $self->enter_mode( $chip->MODE )
      ->then( sub {
         my ( $mode ) = @_;
         $chip->mount( $mode )
      })
      ->then_done( $chip );
}

=head2 $pirate->start->get

Starts binary IO mode on the F<Bus Pirate> device, enabling the module to
actually communicate with it. Normally it is not necessary to call this method
explicitly as it will be done by the setup code of the mode object.

=cut

sub start
{
   my $self = shift;

   Future->wait_any(
      $self->read( 5 )->then( sub {
         my ( $buf ) = @_;
         return Future->done( ( $self->{version} ) = $buf =~ m/^BBIO(\d)/ );
      }),
      repeat {
         $self->write( "\0" );
         $self->sleep( 0.05 );
      } foreach => [ 1 .. 20 ],
        otherwise => sub {
           Future->fail( "Timed out waiting for device to enter bitbang mode" )
        },
   );
}

=head2 $pirate->stop

Stops binary IO mode on the F<Bus Pirate> device and returns it to user
terminal mode. It may be polite to perform this at the end of a program to
return it to a mode that a user can interact with normally on a terminal.

=cut

sub stop
{
   my $self = shift;

   $self->write( "\0\x0f" );
}

# Future support
sub _new_future
{
   my $self = shift;
   return Device::BusPirate::_Future->new( $self );
}

package Device::BusPirate::_Future;
use base qw( Future );
use Carp;

use Time::HiRes qw( time );

sub new
{
   my $proto = shift;
   my $self = $proto->SUPER::new;
   $self->{bp} = ref $proto ? $proto->{bp} : shift;
   return $self;
}

sub await
{
   my $bp = shift->{bp};

   my $alarm = $bp->{alarms}[0];
   my $read = $bp->{read_f}[0];

   my $fh = $bp->{fh};

   croak "Cannot await with nothing to do" unless $alarm or $read;

   my $buf = '';

   while(1) {
      if( $read and length $buf >= $read->[0] ) {
         printf STDERR "PIRATE << %v02x\n", $buf if Device::BusPirate::PIRATE_DEBUG;
         shift @{ $bp->{read_f} };
         $read->[1]->done( substr $buf, 0, $read->[0], "" );
         return;
      }

      my $timeout = $alarm ? $alarm->[0] - time() : undef;
      my $rvec = '';
      vec( $rvec, $fh->fileno, 1 ) = 1 if $read;

      if( select( $rvec, undef, undef, $timeout ) ) {
         $fh->sysread( $buf, $read->[0] - length $buf, length $buf );
      }
      elsif( $timeout ) {
         shift @{ $bp->{alarms} };
         $alarm->[1]->done;
         return;
      }
   }
}

=head1 TODO

=over 4

=item *

More modes - I2C, UART, 1-wire, raw-wire

=item *

Documentation/examples/actual implementations of truely-async subclass.

=item *

PWM, AUX frequency measurement and ADC support.

=back

=head1 AUTHOR

Paul Evans <leonerd@leonerd.org.uk>

=cut

0x55AA;
